File: app/domain/progress.py
"""Playback progress math, extracted from StreamEngine.playback_progress.

Pure function of (state, now): `now` is a monotonic clock reading supplied by
the caller so tests never wait or sleep.
"""

from __future__ import annotations

from typing import Any

from app.domain.playback_state import PlaybackMode, PlaybackState


def playback_progress(state: PlaybackState, now_monotonic: float) -> dict[str, Any]:
    if state.mode != PlaybackMode.playing:
        return {
            "duration_seconds": state.now_playing_duration_seconds,
            "started_at": state.started_at_epoch_seconds,
            "elapsed_seconds": None,
            "progress_percent": None,
        }

    duration = state.now_playing_duration_seconds
    if state.paused and state.paused_elapsed_seconds is not None:
        elapsed_seconds: float | None = state.paused_elapsed_seconds
    elif state.started_at_monotonic_seconds is not None:
        elapsed_seconds = max(0.0, now_monotonic - state.started_at_monotonic_seconds)
    else:
        elapsed_seconds = None

    progress_percent: float | None = None
    if elapsed_seconds is not None and duration:
        progress_percent = min(100.0, elapsed_seconds / duration * 100.0)
    return {
        "duration_seconds": duration,
        "started_at": state.started_at_epoch_seconds,
        "elapsed_seconds": elapsed_seconds,
        "progress_percent": progress_percent,
    }

File: packages/domain/src/progress.ts
/**
 * Playback progress math. Ported from app/domain/progress.py.
 * Pure function of (state, now) — callers inject the clock.
 */

import type { PlaybackModeValue, PlaybackState } from "./playback-state.js";
import { PLAYBACK_MODE_PLAYING } from "./playback-state.js";

export interface PlaybackProgress {
  durationSeconds: number | null;
  startedAt: number | null;
  elapsedSeconds: number | null;
  progressPercent: number | null;
}

export function playbackProgress(state: PlaybackState, nowMonotonic: number): PlaybackProgress {
  if (state.mode !== PLAYBACK_MODE_PLAYING) {
    return {
      durationSeconds: state.nowPlayingDurationSeconds,
      startedAt: state.startedAtEpochSeconds,
      elapsedSeconds: null,
      progressPercent: null,
    };
  }

  const duration = state.nowPlayingDurationSeconds;
  let elapsedSeconds: number | null;
  if (state.paused && state.pausedElapsedSeconds !== null) {
    elapsedSeconds = state.pausedElapsedSeconds;
  } else if (state.startedAtMonotonicSeconds !== null) {
    elapsedSeconds = Math.max(0, nowMonotonic - state.startedAtMonotonicSeconds);
  } else {
    elapsedSeconds = null;
  }

  let progressPercent: number | null = null;
  if (elapsedSeconds !== null && duration && duration > 0) {
    progressPercent = Math.min(100, (elapsedSeconds / duration) * 100);
  }
  return {
    durationSeconds: duration,
    startedAt: state.startedAtEpochSeconds,
    elapsedSeconds,
    progressPercent,
  };
}

export type { PlaybackModeValue };
